### PR TITLE
feat(files): Import/export files to textile bucket AP-799

### PR DESCRIPTION
**Pull request by josephmcg**
**What this PR does** 📖
- Upload/delete files in 'personal-files' textile bucket.
- After file/folder update, upload a JSON file to bucket which keeps track of directory structure
  - this is done with a recursive import & export. May encounter performance issues eventually, but maybe warp will ease this concern
- context menu allowing for like/share/delete of items in grid view
  - rename is currently todo
  - sharing folder is currently todo, toast will indicate that
  - I made a [ticket](https://satellite-im.atlassian.net/browse/AP-828) about this, but the context menu will need a rework to be more dynamic (showing Favorite/un-favorite depending on current file status)
- filesystem.hasChild was only checking the current directory. Now it will check a flattened list of all files which is recursively generated
  - let me know if this algorithm can be optimized

**Which issue(s) this PR fixes** 🔨
AP-799

**Special notes for reviewers** 🗒️
- While the directory structure index is being updated, all the files UI will be disabled. If not, it allows for another update request before the current request finishes, which throws an error. We can update the loading ui if it makes the app feel slow
- I want to improve our nsfw check later. I made a [ticket](https://satellite-im.atlassian.net/browse/AP-807?) about it already, but it will currently reject certain image types as nsfw if they're unable to be scanned (tiff).

**Additional comments** 🎤
- Ticket mentions removing multiple files at once, this is still todo
- if you delete a folder which has child files, the files will still exist in the bucket. the reference however will be deleted from the index. maybe we should delete everything nested inside, would need more recursion

### CODE REVIEW

File: components/views/files/controls/Controls.vue [hunk 1]
```diff
@@ -63,38 +71,49 @@ export default Vue.extend({
...
-            format: 'JPEG',
-            quality: 1,
-          })
-          const fileJpg = new File([outputBuffer.buffer], file.name, {
-            type: 'image/jpeg',
-          })
-          return { file, nsfw: await this.$Security.isNSFW(fileJpg) }
-        }
+      this.errors = []
+      this.$store.commit('ui/setIsLoadingFileIndex', true)
```
[the remark this answers is not in the document]
Comment (josephmcg) [reply components/views/files/controls/Controls.vue:75]: Thanks, I noticed this as well.. probably going to update it to use data and pass props/emit instead.

[the remark this answers is not in the document]
Comment (josephmcg) [reply components/views/files/controls/Controls.vue:75]: Actually I'll just change the persist config. I think keeping in store is cleaner than lots of props/emits

**Comment by phillsatellite** [comment]
Tested: Worked good only issue I came across was I would get a crash whenever trying to upload .heic images
Screen Shot 2022-02-24 at 4 40 41 PM

**Review by WanderingHogan — APPROVED** [review]

**Review by phillsatellite — APPROVED** [review]

**Comment by stavares843** [comment]
Captura de ecrã 2022-02-25, às 00 52 57

not sure if this is already on your list for other PR but this view is missing a scroll if you upload too many files

also a couple of questions
- are we gonna have a limit of file size to upload? I tried with 100MB and worked
- are we gonna have a limit on the number of files to upload at the same time? I tried 9 and worked

I noticed we don't allow uploading a file with the same name even on different folders, is that on purpose?

https://user-images.githubusercontent.com/29093946/155633110-414894ff-e2ec-4800-b74d-5f9c421cf34e.mov

also if you have an error and you switch folders, the error is still there

Captura de ecrã 2022-02-25, às 00 57 30

**Comment by stavares843** [comment]
merging, all of the above can be addressed separately

**Comment by josephmcg** [comment]
Thanks Sara

- this was focused mostly on grid view, I'll make a ticket for list view scroll.
- idk about file size limit or file number limit. I recall Matt saying 8gb, but that figure may have changed.
- Yes, disallowing the same name in different folders was intentional. The folders are for organization on the client end. They don't exist in the textile bucket, all the files are at the root. If we use the same name, it would overwrite other files
- we can clear errors on a folder change if you like. we can consult ui to see if that's better

**Pull request merged**